---
title: "Construct CLL data objects used in the MOFAtools vignette"
author: "Britta Velten"
date: "9/18/2017"
output:
  BiocStyle::html_document:
    toc: true
---

```{r, message=F, warning=FALSE}
library(dplyr)
library(DESeq2)
library(data.table)
library(PACEdata)
```

In this script the pre-processing steps to generate the data objects CLL_data.RData and CLL_covariates.RData stored in the MOFAtools package.
Starting point is the data contained in the 'PACEdata' pacakge based on Dietrich, Oles, Lu et al.

# Set general in/out parameters
```{r}
# minimum number of views observed for a patient to be included out of 4 (RNA, meth, viab, mut), covariates are not considered
minViews <- 2

outDir <- paste("CLLdata_minView=",minViews, sep="")
if(minViews=="all") minViews <- 4
if(!dir.exists(outDir))  dir.create(outDir)

#patients to be included (only CLL)
data("patmeta", package = "PACEdata")
patmeta$PatID<-rownames(patmeta)
patCLL<-patmeta %>% filter(Diagnosis=="CLL")
rownames(patCLL)<-patCLL$PatID
```

# Get RNAseq
```{r}
source("getRNAseq.R")
  
# annotation files downloaded from ENSEMBLE (corresponding to GRCh 37.1)
AnnoFileList<-list(
  mRNA= "Hsapiens_genes_BioMart.75.txt",
  lincRNA = "Hsapiens_lincRNA_BioMart.75.txt",
  miRNA = "Hsapiens_miRNA_BioMart.75.txt"
)

RNA <- getRNAseq(pat2include = patCLL$PatID,
                minrs = 100, AnnoFileList = AnnoFileList, plotit = T,
                outdir = outDir, onlyCD19 = F, topmRNA=5000, noY=T)

any(is.na(Reduce(cbind,RNA)))
```

# Get methylation data
```{r}
source("getMeth.R")
# met.tsv.gz contains chromosome annotation for each CpG site
meth <- getMeth(methFile, pat2include = patCLL$PatID,
              Frac2include = 0.01, outdir = outDir, includeXYchr=F,
              methDataFile="zcat < met.tsv.gz")
any(is.na(meth))
```

# Get drug response 
```{r}
source("getViab.R")
# Threshold parameters: drugs are accepted that for at least `effectNum` samples 
# have a viability effect less than or equal to `effectVal`. On the other hand, the 
# mean viability effect should not be below `viab`.

# we include all drugs and conc
viab <- getViab(pat2include = patCLL$PatID,
                  badDrugs=c( "D_008",  "D_025"), 
                  conc2include = 1:5,
                  targetedDrugs= c("D_002", "D_003", "D_166", "D_082", "D_079", "D_012", "D_030", "D_063", "D_083") , 
                  conc4targeted = 1:5,
                  chemoDrugs = c("D_006", "D_159", "D_010"),
                  conc4chemo = 1:5,
                  effectNum = 1,
                  effectVal = 2,
                  viab = 0, 
                  maxval = 1.1,
                  plotit =F,
                  outdir = outDir)
any(is.na(viab))
```

# Get mutations
```{r}
source("getMut.R")
mut<-getMut(pat2include = patCLL$PatID,
                  minObs = 3,
                  outdir = outDir)

#add IGHV status to mutations
ighv<-patCLL[rownames(mut),"IGHV"]
ighv<-ifelse(ighv=="U",0,1)
mut<-cbind(mut, IGHV=ighv)

any(is.na(mut))
any(apply(mut,1,function(r) all(is.na(r))))
```

# Get covariates
```{r}
source("getCovariates.R")
covariates <- getCovariates(pat2include = patCLL$PatID,
              cov2include = c("Gender", "Diagnosis"),
              outdir = outDir)
```

# Overview Samples per Omic
```{r}
PatPerView<-lapply(list(RNA=RNA[[1]], meth=meth, viab=viab, mut=mut, covariates = covariates), rownames)
gplots::venn(PatPerView)
```

# Save
## Form union of all patients
```{r}
unionPats <- Reduce(union, PatPerView)
subset_augment<-function(mat, pats) {
  aug_mat<-matrix(NA, ncol=ncol(mat), nrow=length(pats))
  aug_mat<-mat[match(pats,rownames(mat)),,drop=FALSE]
  rownames(aug_mat)<-pats
  colnames(aug_mat)<-colnames(mat)
  return(aug_mat)
}
viab_aug<-subset_augment(viab, pats=unionPats)
meth_aug<-subset_augment(meth, pats=unionPats)
mut_aug<-subset_augment(mut, pats=unionPats)
RNA_aug<-lapply(RNA, function(RNAsub) subset_augment(RNAsub, pats=unionPats))
covariates_aug <- subset_augment(covariates, pats=unionPats)

CLL_aug <- list(mRNA = RNA_aug$mRNA, viab = viab_aug, meth= meth_aug, mut =mut_aug)
observedViewsPerPat <- rowSums(sapply(CLL_aug, function(dat) !apply(is.na(dat),1,all)))
patsWithminViews <- observedViewsPerPat >= minViews
patsWithminViews <- names(patsWithminViews)[patsWithminViews]
```

## Save views with samples having at least minViews
```{r}
  write.table(viab_aug[patsWithminViews,], file=file.path(outDir,"viab.txt"),
                row.names=TRUE, col.names=TRUE, quote=F)
  write.table(meth_aug[patsWithminViews,], file=file.path(outDir,"meth.txt"),
                row.names=TRUE, col.names=TRUE, quote=F)
  write.table(mut_aug[patsWithminViews,], file=file.path(outDir,"mut.txt"),
                row.names=TRUE, col.names=TRUE, quote=F)
  write.table(covariates_aug[patsWithminViews,], file=file.path(outDir,"covariates.txt"),
                row.names=TRUE, col.names=TRUE, quote=F)
  for(nm in names(RNA_aug)) write.table(RNA_aug[[nm]][patsWithminViews,],
                                        file=file.path(outDir,paste(nm,".txt", sep="")),
                                        row.names=TRUE, col.names=TRUE, quote=F)
```

## Save objects used for vignette
```{r}
CLL_data <- list(Methylation = meth_aug[patsWithminViews,],
                 mRNA = RNA_aug[["mRNA"]][patsWithminViews,],
                 Drugs = viab_aug[patsWithminViews,],
                 Mutations = mut_aug[patsWithminViews,])

sapply(CLL_data, dim)
save(CLL_data, file="CLL_data.RData")


save(covariates_aug[patsWithminViews,], file="CLL_covariates.RData")
```
